### PR TITLE
Add Pivot AllToAll algorithm for Rome model

### PR DESCRIPTION
**Pull request by yzygitzh**
This commit adds a new Pivot AlltoAll algorithm for Rome model. The algorithm is bandwidth-optimal for ring-based topologies, and brings ~70% speedup compared with direct P2P based AllToAll on rome_model_56.

### CODE REVIEW

File: src/collectives/device/common.h [hunk 1]
```diff
@@ -618,4 +622,9 @@ __device__  __attribute__((noinline)) void NCCL_FUNC_NAME(func, algo, proto, dev
   IMPL_COLL_FUNC(func, RING, SIMPLE, Sum, int8_t); \
   IMPL_COLL_KERN(func, RING, SIMPLE, Sum, int8_t, FUNC_INDEX_P2P);
 
+// AllToAll Pivot primitive only has one function/kernel.
+#define IMPL_COLL_ALLTOALL_PIVOT(func) \
+  IMPL_COLL_FUNC(func, RING, SIMPLE, Sum, int8_t); \
+  IMPL_COLL_KERN(func, RING, SIMPLE, Sum, int8_t, FUNC_INDEX_ALLTOALL_PIVOT);
```
Comment (wenkaidu) [inline src/collectives/device/common.h:628]: You can remove IMPL_COLL_KERN. Only kernel instance is SendRecv.

File: src/enqueue.cc [hunk 2]
```diff
@@ -423,7 +423,11 @@ static ncclResult_t getAlgoInfo(struct ncclInfo* info, int collNetTypeSupport, i
     if (info->algorithm == NCCL_ALGO_COLLNET) nt += 3*WARP_SIZE;
   }
 #endif
-  info->nChannels = nc;
+  if (info->coll == ncclFuncAllToAllPivot) {
+    info->nChannels = comm->topo->pivotA2ANumChannels;
```
Comment (wenkaidu) [inline src/enqueue.cc:427]: We can only use up to established channel, thus need to take std::min from pivotA2ANumChannels and nc.

File: src/collectives/all_to_all_api.cc [hunk 3]
```diff
@@ -7,20 +7,28 @@
...
   ncclComm_t comm, hipStream_t stream) {
-  int nRanks;
-  NCCLCHECK(ncclCommCount(comm, &nRanks));
-  size_t rankOffset = count * ncclTypeSize(datatype);
-  if (count == 0) return ncclSuccess;
-  NCCLCHECK(ncclGroupStart());
-  for (int r=0; r<nRanks; r++) {
-    NCCLCHECK(ncclSend(((char*)sendbuff)+r*rankOffset, count, datatype, r, comm, stream));
-    NCCLCHECK(ncclRecv(((char*)recvbuff)+r*rankOffset, count, datatype, r, comm, stream));
+  if (comm->topo->pivotA2AEnabled) {
```
Comment (wenkaidu) [inline src/collectives/all_to_all_api.cc:16]: Decision to take pivot alltoall path needs to be "voted" using https://github.com/ROCmSoftwarePlatform/rccl/blob/develop/src/init.cc#L1044. This is to make sure all participating ranks are in sync. You can add a field in allGather3Data.

File: src/collectives/device/common.h (hunk 1 above)
Comment (yzygitzh) [reply src/collectives/device/common.h]: Fixed.

File: src/enqueue.cc (hunk 2 above)
Comment (yzygitzh) [reply src/enqueue.cc]: Fixed.

File: src/enqueue.cc [hunk 4]
```diff
@@ -760,7 +773,11 @@ ncclResult_t ncclSetupAsyncKernels(ncclComm_t comm) {
     int allCollNetSupport = comm->collNetSupport;
     for (int c = 0; c < comm->asyncOpCount; c++) {
       struct ncclInfo* info = comm->asyncOps+c;
-      info->nChannels = std::min(std::max(1, (int)DIVUP(info->nBytes, channelSize)), comm->nChannels); // assign number of channels
+      if (info->coll == ncclFuncAllToAllPivot) {
+        info->nChannels = comm->topo->pivotA2ANumChannels;
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply src/enqueue.cc]: Fixed.

File: src/collectives/all_to_all_api.cc (hunk 3 above)
Comment (yzygitzh) [reply src/collectives/all_to_all_api.cc:16]: Fixed.

**Review by wenkaidu — APPROVED** [review]

**Pull request merged**